examples/chip-tool/commands/interactive/InteractiveCommands.cpp

```diff
@@ -378,10 +388,67 @@ CHIP_ERROR InteractiveStartCommand::RunCommand()
         command = nullptr;
     }
 
+    JoinCommandExecutorThread();
+    CHIPCommand::RemoveExtraCheckInDelegate(this);
     SetCommandExitStatus(CHIP_NO_ERROR);
     return CHIP_NO_ERROR;
 }
 
+void InteractiveCommand::StartCommandExecutorThread()
+{
+    commandExecutorThread = std::thread(&InteractiveCommand::CommandExecutor, this);
+}
+
+void InteractiveCommand::JoinCommandExecutorThread()
+{
+    {
+        std::lock_guard<std::mutex> lock(commandExecutorMutex);
+        commandExecutorQueue.push(CommandExecutorTask(false));
+        commandExecutorCv.notify_all();
+    }
+    commandExecutorThread.join();
+}
+
+void InteractiveCommand::CommandExecutor()
+{
+    std::vector<CommandExecutorTask> tasks;
+    while (true)
+    {
+        // Hold the lock as short as possible. Since OnCheckInComplete also uses the lock.
+        // If the lock covers execution call it will be a deadlock when a check-in message arrived during executing the commands.
+        {
+            std::unique_lock lock(commandExecutorMutex);
+            commandExecutorCv.wait(lock, [this]() { return !commandExecutorQueue.empty(); });
+
+            while (!commandExecutorQueue.empty())
+            {
+                tasks.emplace_back(commandExecutorQueue.front());
+                commandExecutorQueue.pop();
+            }
+        }
+
+        for (const auto & task : tasks)
+        {
+            switch (task.kind)
+            {
+            case CommandExecutorTask::Kind::STOP:
+                return;
+            case CommandExecutorTask::Kind::ON_CHECK_IN_COMPLETE:
+                mHandler->RunAllQueuedCommandsForNode(task.payload.Get<chip::ScopedNodeId>(), GetStorageDirectory(),
```

Comment:
This is calling into the "run a command" code on this executor thread?

But the user can also call into that code, racing this call, by typing and hitting enter?

That code is not threadsafe and this will break.

Reply:
Added a mutex here.

I'm also thinking to send CLI input to this thread to execute (in future PRs).